pkg/controllers/jobset_controller.go

```diff
@@ -390,6 +390,7 @@ func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, js *
 				Name:      subdomain,
 				Namespace: js.Namespace,
 			},
+			// TODO how to handle replicated job selectors here when more than one job share a service?
```

Comment:
hmm, this is indeed problematic. The `serviceSpec.selector` doesn't allow building an OR label selector, and so all the jobs must have a common label for this to work.

So, instead of allowing to set a subdomain at the replicatedJob level, how about allowing to create a shared service at the jobset level using the name of the jobset? If one wants to setup a single service across multiple JobSets, then they will need to do that manually.

Reply:
This implies adding a Network->EnableDNSHostnames field to the JobSetSpec struct.

Reply:
What about, when a subdomain is selected we allow the user to define a custom SelectorLabel that (if defined) is used / applied to the replicated jobs in the jobsets, and used by the service? And if the SelectorLabel isn't supplied, we create one that is based on the subdomain instead.

> So, instead of allowing to set a subdomain at the replicatedJob level, how about allowing to create a shared service at the jobset level using the name of the jobset?

That could work - that would essentially be what we are doing, but instead of allowing a subdomain, just having some kind of boolean for a shared service at the level of the JobSet? That would work and just we would need to ensure that if they defined something under some JobSet -> Network they don't also have something that might conflict under ReplicatedJob -> Network. Arguably, we could still honor the `EnableDNSHostnames` but it would be using the JobSet level service.

> If one wants to setup a single service across multiple JobSets, then they will need to do that manually.

That would be fair - and I think in this case it would look like what I'm currently doing with the operator.

So what do you think - should we have something like:

```go
// JobSetSpec defines the desired state of JobSet
type JobSetSpec struct {
	// ReplicatedJobs is the group of jobs that will form the set.
	// +listType=map
	// +listMapKey=name
	ReplicatedJobs []ReplicatedJob `json:"replicatedJobs,omitempty"`

	// Network on the level of the Jobset
	Network JobSetNetwork `json:"network,omitempty"`
...
}

type JobSetNetwork struct {
	// Subdomain is an explicit choice for a network subdomain name
	// When set, the all replicated jobs with DNS enabled are added to this network.
	// +optional
	Subdomain string `json:"subdomain,omitempty"`
}
```

The reason I left subdomain is so the user could more explicitly change the name of the network, and then maybe be able to apply that predictable label elsewhere. Also, we would want to use that as a flag that it is provided / enabled. If we don't have subdomain (and default to an attribute of the jobset like the name) we would need to have something kind of like:

```go
type JobSetNetwork struct {
	// Enable the network on the level of the JobSet for Replicated jobs that have DNS enabled
	// +optional
	Enabled bool `json:"enabled,omitempty"`
}
```
 
or of course I am open to other designs! But I do think I like the idea of moving this up one level, and using the JobSet as the identifier for things. Let me know what you think of the above.

Reply:
I think we could have the `JobSetNetwork.Subdomain` as a field the user can set, and just add defaulting to set it to `<jobSetName>-<replicatedJobName>` (how we currently set it). @ahg-g what do you think?

Reply:
That would work - I'll wait to hear from @ahg-g before taking a next step!

Reply:
I suggest we *move* the network struct from replicatedJob to JobSetSpec. In the future, if we decided we want to configure which exact replicatedJobs should have dns hostnames enabled, we can add a target field like we did for success policy. The subdomain will be the JobSet name.

Reply:
Moving will require us to upgrade the api to v1alpha2

Reply:
@ahg-g would you like me to do those changes in this PR too?

Reply:
Perhaps doing it over two PRs is better, the first to bump v1alpha1 to v1alpha2 and the second to apply the changes of moving the struct and updating the logic in the operator.

Reply:
Working on step two now - no need to test the most recent push.